### PR TITLE
ZEPPELIN-142 add SparkSqlCompleter

### PR DESCRIPTION
**Pull request by nevillelyh**

### CODE REVIEW

File: spark/src/main/java/org/apache/zeppelin/spark/SparkSqlInterpreter.java [hunk 1]
```diff
@@ -292,6 +293,7 @@ public Scheduler getScheduler() {
 
   @Override
   public List<String> completion(String buf, int cursor) {
-    return null;
+    SQLContext sqlc = getSparkInterpreter().getSQLContext();
+    return new SparkSqlCompleter(sqlc).completion(buf, cursor);
```
Comment (felixcheung) [inline spark/src/main/java/org/apache/zeppelin/spark/SparkSqlInterpreter.java]: isn't this going to create a new instance of SparkSqlCompleter each time code completion is invoked on a text block?

File: spark/src/main/scala/org/apache/zeppelin/spark/utils/SparkSqlCompleter.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+class SparkSqlCompleter(val sqlContext: SQLContext) {
+
+  private val keywords = Source.fromInputStream(this.getClass.getResourceAsStream("/sparksql.txt")).getLines().toSet
+
+  def completion(buf: String, cursor: Int): java.util.List[String] = {
+    import sqlContext.implicits._
+
+    val tables = sqlContext.tables().select('tableName).collect().map(_.getString(0))
+    val fields = tables.flatMap(getFields)
```
Comment (felixcheung) [inline spark/src/main/scala/org/apache/zeppelin/spark/utils/SparkSqlCompleter.scala]: how expansive is it to call collect() or table.schema.field on sqlContext for every call to completion?
is there a way to cache it some how?

Comment (nevillelyh) [inline spark/src/main/scala/org/apache/zeppelin/spark/utils/SparkSqlCompleter.scala]: Added Guava suppliers for list of tables and cache for fields in tables.

File: spark/src/main/java/org/apache/zeppelin/spark/SparkSqlInterpreter.java (hunk 1 above)
Comment (nevillelyh) [inline spark/src/main/java/org/apache/zeppelin/spark/SparkSqlInterpreter.java]: Moved SparkSqlCompleter construction to `open()`

**Comment by nevillelyh** [comment]
Looks like this failed for spark-sql < 1.3. What's the recommended way of dealing with multiple Spark API versions?

**Comment by nberserk** [comment]
when this completion invoked? I wonder.

**Comment by nevillelyh** [comment]
@nberserk you invoke it with `ctrl-.` in a `%sql` note.

**Comment by Leemoonsoo** [comment]
Thanks for the great contribution.

According to message from CI server

```
[ERROR] /home/travis/build/apache/incubator-zeppelin/spark/src/main/scala/org/apache/zeppelin/spark/utils/SparkSqlCompleter.scala:8: error: object StructType is not a member of package org.apache.spark.sql.types
[INFO] import org.apache.spark.sql.types.{StructType, StructField}
[INFO]        ^
[ERROR] /home/travis/build/apache/incubator-zeppelin/spark/src/main/scala/org/apache/zeppelin/spark/utils/SparkSqlCompleter.scala:15: error: value implicits is not a member of org.apache.spark.sql.SQLContext
[INFO]   import sqlContext.implicits._
[INFO]                     ^
[ERROR] /home/travis/build/apache/incubator-zeppelin/spark/src/main/scala/org/apache/zeppelin/spark/utils/SparkSqlCompleter.scala:20: error: value tables is not a member of org.apache.spark.sql.SQLContext
[INFO]       override def get(): Seq[String] = sqlContext.tables().select('tableName).collect().map(_.getString(0))
[INFO]                                                    ^
[ERROR] /home/travis/build/apache/incubator-zeppelin/spark/src/main/scala/org/apache/zeppelin/spark/utils/SparkSqlCompleter.scala:40: error: not found: type StructField
[INFO]     def fieldNames(prefix: Seq[String], fields: Array[StructField]): Seq[String] = {
[INFO]                                                       ^
[ERROR] /home/travis/build/apache/incubator-zeppelin/spark/src/main/scala/org/apache/zeppelin/spark/utils/SparkSqlCompleter.scala:41: error: type mismatch;
[ERROR]  found   : fields.type (with underlying type Array[<error>])
[INFO]  required: ?{def flatMap: ?}
[INFO] Note that implicit conversions are not applicable because they are ambiguous:
[INFO]  both method booleanArrayOps in object Predef of type (xs: Array[Boolean])scala.collection.mutable.ArrayOps[Boolean]
[INFO]  and method byteArrayOps in object Predef of type (xs: Array[Byte])scala.collection.mutable.ArrayOps[Byte]
[INFO]  are possible conversion functions from fields.type to ?{def flatMap: ?}
[INFO]       fields.flatMap { f =>
```

It looks like there're some compiling error. @nevillelyh Do you have any idea about the error?

**Comment by nevillelyh** [comment]
Yeah I'm using API introduces in Spark SQL 1.3. What's the best way to make this work? Separate src/man/java directories for each spark version + Maven profiles?

**Comment by Leemoonsoo** [comment]
Can we go with reflection?

for example, api of getting progress in spark vary between versions. And it is handled using reflections, like https://github.com/apache/incubator-zeppelin/blob/branch-0.5/spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java#L653

What do you think?

**Comment by nevillelyh** [comment]
It might be hard and unreadable since I rely on types `StructField` and `StructType` being available in the compiler.

**Comment by Leemoonsoo** [comment]
Hmm. I'm worried about separating src directory per spark version will make things too complicated.
And in the future, Zeppelin might have ability to work with user provided spark dependency. (like spark has package for user provided hadoop). It'll be difficult to support such case when codes are separated per spark version.

What do you think? Which one do you think better?

**Comment by nevillelyh** [comment]
Yeah make sense. I'll try to refactor it and remove the pattern matching parts.

**Comment by corneadoug** [comment]
@nevillelyh 
I'm going around old PRs, is there any news on this issue?
Or do you want to close it?

**Comment by nevillelyh** [comment]
Sorry haven't worked on this in a while and probably won't have time to. Closing it for now.

**Pull request closed without merging**